### ISSUE TITLE
Include benchmark extension

### ISSUE THREAD

**Issue by nipafx** [junit-pioneer/junit-pioneer#99]
[My JUnit 5 demo](https://github.com/CodeFX-org/demo-junit-5) contains [a benchmark extension](https://github.com/CodeFX-org/demo-junit-5/blob/master/src/main/java/org/codefx/demo/junit5/BenchmarkExtension.java) that very naively uses `System.currentTimeMillis()` to report a test's wall clock time. Use:

```java
@Test
@Benchmark
void benchmarked() throws InterruptedException {
	Thread.sleep(100);
	assertTrue(true);
}
```

Output:

```
timestamp = 2018-07-28T09:53:39.128030, Benchmark = Test 'benchmarked()' took 101 ms.
timestamp = 2018-07-28T09:53:39.136931, Benchmark = Test container 'BenchmarkTest' took 152 ms
```

I think this is worthy for inclusion in JUnit Pioneer. Thoughts?

**Comment by nicolaiparlog** [comment]
Possible improvement:

* register extension globally
* activate for all tests if configuration property is set

**Comment by nicolaiparlog** [comment]
Possible improvement:

* aggregate and output data after test suite ran
* e.g. show x longest running tests / containers

**Comment by fduminy** [comment]
Maybe you should use ```java.time.Clock``` and its subclasses (I also found one from junit5 : ```org.junit.platform.console.tasks.IncrementingClock```) or Guava's [Stopwatch](https://google.github.io/guava/releases/snapshot-jre/api/docs/com/google/common/base/Stopwatch.html) and [Ticker](https://google.github.io/guava/releases/snapshot-jre/api/docs/com/google/common/base/Ticker.html)

**Comment by fduminy** [comment]
That sound also related to https://github.com/junit-pioneer/junit-pioneer/issues/38

**Comment by nicolaiparlog** [comment]
@fduminy Thank you for the tips, I didn't know about `Stopwatch`. :+1:

**Comment by Bukama** [comment]
Aside from elapsed time another interesting thing, at least for some tests, could be needed memory. If you consider that, then `@Benchmark` is not a very intuitive annotation for "only" meassuring the needed execution time. But maybe I'm too arrow-minded about the name.

**Comment by Bukama** [comment]
> That sounds also related to #38

I was thinking about migrating one or both of them but I couldn't decide which one because of the following thoughts:

1. Copyrights: The `Stopwatch` is an example by JUnit - so would they allow it to be implemented in JUnitExtension (*wave* @marcphilipp  , @sormuras ). The `Benchmark` one is written by @nicolaiparlog 
2. The purpose: Do we only want to measure time or do we perspectively also measure other things like memory (see my comment from 2019-12-17).

To 1) As the `Benchmark` is written by @nicolaiparlog and he also suggest to include it into pioneer we should have no problems about that, when going for his one (Call this a vote for this).

To 2) For me this is close connected to how the extension should be named in pioneer. At the moment both only handle time. If we maybe want to measure other things like memory then we should go for the `Benchmark` one as the naming is more general but still clearly realted to measure (technical) performance. If we only want to stay to measure time then I like `Stopwatch` more, as the metapher of a stopwatch is clearly connected to measure the elapsed time for a certain action (like a 100m sprint in sports).

If we don't want to decide on the future purpose of the extension now, though I would argue for naming the extension `Benchmark` because we are then "future safe". This said I vote for closing #38 because I see it as a duplicate to this issue. 

We should then also discuss the possible improvements by @nicolaiparlog mentioned in Juliy 2018 above. Aside from that I would argue that the usage of `java.time.Clock` (as suggested by @fduminy ) is already decided by ne number of upvotes.

What are your opinions about that?

**Comment by nicolaiparlog** [comment]
Thanks for the write-up @Bukama! :+1: Your thoughts about naming make sense. I'm not sure whether measuring memory consumption is reliable and without overhead, so let's not commit to that. Instead, let's call this `@Stopwatch` and focus on that use case. If we later want to measure memory consumptiom (or whatever) we can do so in dedicated extensions and then maybe combine them in `@Benchmark`.

For now, let's focus on `@Stopwatch` :

* let's use `java.time.Clock` or another JDK-based solution (no additional dependencies!)
* I think the format I initially picked makes sense, but it can be changed if something else appears to be better
* not sure about the name - would prefer one that doesn't clash with something as ubiquitous as Guava; I'm open for suggestions

Stretch goals:

* register extension globally, activate for all tests if configuration property is set
* aggregate and output data after test suite ran, e.g. show x longest running tests / containers

**Comment by nipafx** [comment]
Just to clarify (or retcon), the stopwatch part will be implemented in #38. If there are any other performance-y things we want to measure, that would be separate issues. This one tracks such extensions and can be used to activate them with a single annotation.

**Comment by nipafx** [comment]
To not have this issue open forever, we decided to close it. If other performance-related extensions show up, we can always reopen it.
